chore: Design doc for the TssBootstrap (#2457)
hiero-ledger/hiero-block-node: docs/design/tss-bootstrap-plugin.md

    - Plugin implements the existing BlockNodePlugin interfaces
    - The planned plugin queries a peer BN's serverStatusDetail gRPC endpoint to retrieve TSS information
      - gRPC peer communication follows the same WebClient pattern used elsewhere in the codebase
    - The planned plugin will persist the TSS data locally and load it during plugin init() allowing it to be ready when block proofs arrive with TSS Signatures.
    - If TSS data is not present locally, the plugin will check to see if the TSS Data is stored in its config. TssBootstrapPlugin config can be used for both testing and temporary initialization for a Block Node
    - The planned plugin will periodically query its peers for TSS data updates.
  - TSS data is exposed to other plugins
    - The BlockNode plugin interface will be refactored to have an onContextUpdated method that will be called by the BlockNode application when the context is updated.
    - The StatusDetailPlugin will implement ```onContextUpdate()``` and receive TSS data updates via the

## Terms

<dl>
  <dt>Term</dt>
  <dd>A Term is any word or short phrase used in this document that
      requires clear definition and/or explanation.</dd>

> </dl>

Reviewer: Leftover?
It would help to define any useful terms here.

## Entities

### BlockNodeContext

- Will contain the ```TssData``` information and is passed to plugins in the ```BlockNodePlugin.init()``` call.
- Could possibly contain the ```PluginManager``` so that plugins can pass context updates directly to the ```PluginManager```

> - The BlockNodeContext will also be sent to the plugins via the ```BlockNodePlugin.onContextUpdate()``` callback function on the plugins

Reviewer: Slight rewording to clarify that this is separate from the plugin.

Note, also, that it's generally recommended to use a _single_ backquote(`) for code elements that are not multiple lines.

```suggestion
- Carries `TssData` information and is passed to plugins in a `BlockNodePlugin.init()` call.
- The `BlockNodeContext` will also be sent to plugins via a `BlockNodePlugin.onContextUpdate()` call to each plugin.
```

> # Design Document Template

Reviewer 2: Update title

Reviewer: ```suggestion
# TSS Bootstrap Plugin Design
```

    - The `ApplicationStateFacility` interface coordinates `BlockNodeContext` and `TssData` changes between plugins
      - Handles requests to change `TssData`
      - Persists the latest `TssData`
      - Notifies plugins when the `BlockNodeContext` changes
      - Implemented by the `BlockNodeApp`
    - The plugin checks it's config at startup to see if `TSSData` is present.
    - `TssBootstrapPluginConfig` can be used for both testing and temporary initialization for a Block Node
    - The plugin will periodically query its peers for TSS data updates.
  - TSS data is exposed to other plugins
    - The BlockNode plugin interface contains an `onContextUpdate` method that is called by the `ApplicationStateFacility` when the context is updated.
    - The `StatusDetailPlugin` implements `onContextUpdate()` and receives TSS data updates from the `ApplicationStateFacility`

## Terms

<dl>
  <dt>TSS</dt>
  <dd>Threshold Signing Scheme, which consists of two main algorithms: WRAPS and HINTS.</dd>
</dl>
<dl>
  <dt>WRAPS</dt>
  <dd>Proves validity of the HINTS verification key  of the network with a given Ledger ID.</dd>

> </dl>

Reviewer: WRAPS is an acronym of "Weighted Roster Attestation Proof System". We might include that here for completeness.

Author: Updated

<dl>
  <dt>HINTS</dt>
  <dd>Produces a HINTS signature attesting to the claim that majority of node by weight attest to block root hash.
      This HINTS signature is verified using a daily HINTS verification key.</dd>

> </dl>

Reviewer: HINTS is actually hinTS (**hin**ted **T**hreshold **S**ignature scheme), though it's a bit pedantic to harp on...

Author: Updated.

<dl>
  <dt>Verification Key</dt>

>   <dd>The HINTS verification key used to verify HINTS signature on a message, which in our case is the block root hash. This key is 616 bytes.</dd>

Reviewer: This is actually the WRAPS verification key.  The HinTS verification key rotates daily and is part of what the WRAPS proves.

Author: Updated

</dl>
<dl>
  <dt>Ledger ID</dt>
  <dd>SHA-256 hash of the genesis TSS Address Book. This is 32 bytes.</dd>
</dl>
<dl>
  <dt>Schnorr public / private key</dt>
  <dd>used by WRAPS for the purpose of signing the next day’s TSS Address Book and HINTS verification key</dd>

> </dl>

Reviewer: BTW, the Schnorr keys are not used by WRAPS, but by hinTS.
absent the WRAPS, we accept the Schnorr keys directly (and assume equal weight).

Once we have the WRAPS we can use the hinTS key generated/proved in the WRAPS and verify the _aggregate_ signature (which is generally much smaller than the concatenated Schnorr signatures).

The "magic" of BLS is that it takes a set of Schnorr signatures and produces a single signature that only verifies if a sufficient number of valid Schnorr signatures were combined.

What hinTS does is support _weight_ for each signature, so instead of just count the verification requires enough aggregate _weight_.

What WRAPS does is provide a SNARK that proves the roster, which includes the individual public keys, the daily verification key, and the weight for each signer.  That roster is what provides the "hints" needed to validate correctly with respect to _weight_ rather than just _count_.  The daily verification key is what we actually use to verify the BLS signature.

Author: I got this from Notion: https://www.notion.so/swirldslabs/TSS-1f17c9ab2591801d83c5e16fb820df12?source=copy_link#2cd7c9ab2591803e92f6f116a64ae166

Reviewer: It's used in a few cases. All short descriptions are simplifications, but perhaps a more _clear_ simplification might be as follows:
> A key produced privately by each node and used during the address book update process to adopt a new TSS roster. This key may also directly sign a block proof if the WRAPS process is not enabled.

Author: Updated

## Entities

> ### ApplicationStataFacility

Reviewer 2: ```suggestion
### ApplicationStateFacility
```

- Responsible for updating the plugins when context changes.
- Calls the `BlockNodePlugin.onContextUpdate()` on all plugins when the `BlockNodeContext` changes.

> - Passed directly to the plugins as a parameter to the `init()` method.

Reviewer 2: Q: So `BlockNodePlugin.init()` would become `init(BlockNodeContext context, ServiceBuilder serviceBuilder, ApplicationStateFacility applicationStateFacility)`?

Reviewer: The current method is `init(BlockNodeContext context, ServiceBuilder serviceBuilder)` (see [here](https://github.com/hiero-ledger/hiero-block-node/blob/main/block-node/spi-plugins/src/main/java/org/hiero/block/node/spi/BlockNodePlugin.java#L55)).

This effort should add `ApplicationStateFacility` to the `BlockNodeContext` record, not as an added parameter to the `init` method.

Author: I removed ApplicationStateFacility from `init` and added it to the BlockNodeContext. Updated the documentation to match

> - Plugins can use the `updateTssData` method to update the `TssData` in the `BlockNodeContext`

Reviewer 2: To be clear plugins would call `ApplicationStateFacility.updateTssData()` and this is intended to update the instances of `BlockNodeContext` in all other plugins?

Reviewer: Not quite.
`TssBootstrapPlugin` will call `ApplicationStateFacility.updateTssData()` and
the `ApplicationStateFacility` will call `onContextUpdate` for each plugin.

Later (with `VerificationPlugin` updates) the `VerificationPlugin` will call `ApplicationStateFacility.updateTssData()` after extracting TSS data from block `0`; and the same update will propagate that data to other plugins.

In both cases, the task of persisting the data will belong to the `ApplicationStateFacility`, so we have a single plugin managing both storing and loading the data (which avoids multiple plugins needing to manage individual storage and avoids multiple plugins getting too out of sync for the same data).

Author: Joseph's description is accurate.

- Processes requests to change `TssData`
- Persists `TssData`
- Loads persisted `TssData` prior to plugin initialization
- Implemented by the `BlockNodeApp`

### BlockNodeContext

- Carries `TssData` information and is passed to plugins in a `BlockNodePlugin.init()` call.
- The `BlockNodeContext` will also be sent to plugins via a `BlockNodePlugin.onContextUpdate()` call to each plugin.

> ### ServiceStatusServicePlugin

Reviewer 2: ```suggestion
### ServerStatusServicePlugin
```

- The plugin responsible for the `serverStatus`  and `serverStatusDetail` gRPC calls.
- Implements the `BlockNodePlugin.onContextUpdate()` to receive `TssData` updates.
- Responds to peer requests for `TssData` via the `serverStatusDetail` call.

### TssBootstrapConfig

- Used to configure the BN Peers that will be used to query `serverStatusDetail`
- Used to bootstrap TSS data via the Config

> - Overrides initial data provided by the `ApplicationStateFacility`

Reviewer: I wonder if we want this to be the reverse, any data stored in the `ApplicationStateFacility` overrides configuration.
Not sure, both alternatives have downsides.

Author: I did this for testing so that the config data provided by the test will win.

Reviewer: The problem this raises is that _in production_ we could receive an update, then on restart that update gets wiped out because an older value is still set in config.

If nothing ever goes wrong this might be OK, but if the BN is restarted without clearing the config, it could be problematic.

Author: ApplicationStateFacility and all the plugins can now init without getting any context updates. Any TssData updates they submit will be added to a ConcurrentLinkedQueue. After plugin initialization, the ApplicationStateFacility takes the TssData with the greatest validFromBlock and calls the onContextUpdate method for each plugin. So if the config is the greatest, it wins. If not, the persisted data wins.

### TssBootstrapPlugin

- Queries peer BNs for `TssData` via the `serverStatusDetail` gRPC call.
- `TssData` can be configured using the `TssBootstrapConfig`
- `TssBootstrapConfig` data will override the `TssData` received during `init()`.
- Notifies the `ApplicationStateFacility` when it has a `TssData` update.

### VerificationPlugin

- Verifies block 0
- Notifies the `ApplicationStateFacility` when it detects changes to `TssData`
- Implements the `BlockNodePlugin.onContextUpdate()` to receive `TssData` updates

### TssData

- The protobuf `TssData` used in the `serverStatusDetailsResponse`

```markdown
/**
* TSS information for a Block-Node.<br/>
* `TssData` contains the Ledger Id, wraps validation key, and TssRoster information
*
* All fields in this message SHALL be filled in, or none.
*/

>   message TssData {

Reviewer 2: If this is a snippet from CN proto (which I think it is) we should link back to it.

Reviewer: Not CN; this is our own definition and should be in BN API.

Author: Correct. This is part of the BN API.

Reviewer 2: Okay, no-op then.

  /**
    * The ledger id
    */
      bytes ledger_id = 1;
   /**
    * The wraps verification key
    */
      bytes wraps_verification_key = 2;
   /**
    * The current TSS roster
    */
      TssRoster current_roster = 3;
    }

/**
* TSS Roster information.
*
* `TssRoster` contains a list of roster entry
*/
  message TssRoster {
    /**
     * The list of `RosterEntry`
     */
      repeated RosterEntry roster_entries = 1;
      }

/**
* A single TSS Roster entry.
*
* All fields are REQUIRED.<br/>
* The `node_id` field SHALL match the same field for an entry in the Node Store
* in consensus network state.<br/>
* The `schnorr_public_key` MAY be a placeholder value if the associated node failed
* to participate in the roster election within the required time limit.
*/
  message RosterEntry {
  /**
   * The node id
   */
      uint64 node_id = 1;
   /**
    * The node weight
    */
      uint64 weight = 2;
   /**
    * The schnorr public key
    */
      bytes schnorr_public_key = 3;

> }

Reviewer: General note that these should have 4 space indentation and the doc comments should align with the fields, no 1-space extra indent.

There should be a blank line after each field.

Author: corrected

```

## Design

- Oustanding design decisions

>   - Should the TssData or TssRoster messages have the starting block number that the TssData is valid from.

Reviewer 2: I believe so.
It would support the ability to know if the roster is applicable or not

Reviewer: I agree, the TSSData _and_ TSSRoster must _each_ include a `valid_from_block` field (and these might not be the same value).

Author: valid_from_block has been added to both TssData and TssRoster and the design doc has been updated to match.

## Diagram

TBD

Consider using mermaid to generate one or more of the following:
- [User Journey](https://mermaid.js.org/syntax/userJourney.html)
- [Requirement Diagram](https://mermaid.js.org/syntax/requirementDiagram.html)
- [Sequence Diagram](https://mermaid.js.org/syntax/sequenceDiagram.html)
- [Class Diagram](https://mermaid.js.org/syntax/classDiagram.html)
- [Block Diagram](https://mermaid.js.org/syntax/block.html)
- [Architecture Diagram](https://mermaid.js.org/syntax/architecture.html)
- [Flowchart](https://mermaid.js.org/syntax/flowchart.html)
- [State Diagram](https://mermaid.js.org/syntax/stateDiagram.html)
- [Entity Relationship Diagram](https://mermaid.js.org/syntax/entityRelationshipDiagram.html)**

## Configuration

> TBD

Reviewer 2: Please fill this in or note none

Author: Added.

## Metrics

TBD

## Exceptions

TBD

## Acceptance Tests

> TBD

Reviewer 2: Please fill this in

Author: Added

## FAQ

> 1. Plugin ordering — how does bootstrap init before verification? ServiceLoader doesn't guarantee order.

Reviewer 2: Q: Does this mean `TssBootstrapPlugin` must finish its initial logic before before VerificationPlugin needs the data?

Reviewer: Not exactly.
All `init` calls must complete before the block node is "ready".
By the time all `init` calls are completed, the Verification Plugin will have either loaded with correct TSSData or received a Context update with correct TSSData.

So yes, the plugin must finish init before Verification needs the data; but that's already guaranteed by the app.

Author: Some additional clarification. The `ApplicationStateFacility` loads the last persisted `TssData` prior to any `init()` calls. This allows the `TssBootstrapPlugin` to overwrite with `TssData` with TssData stored in the config. The VerificationPlugin could also override the `TssData` with what it thinks is the latest.

Reviewer: I still think that's risky.
I would expect ApplicationStateFacility to load data _during_ init, as plugins really aren't supposed to do much before then (even facility-type plugins).

Config overriding loaded data is difficult as it's risky (we could erase a critical update unexpectedly because config wasn't cleared), but also difficult to prevent (though there are some things we can do in that regard).

Something to consider, perhaps, in future updates.

Reviewer 2: My concern was what if VerificationPlugin init runs first, how do we make sure it gets the latest data since the `TssBootstrapPlugin.init()` doesn't present the latest data till after?
In that case it seems the `ApplicationStateFacility` call in the `TssBootstrapPlugin` will ensure the `VerificationPlugin` `BlockNodeContext` gets updated?

Reviewer: All init happens long before any plugin might have new data.
Data storage belongs to the ApplicationState plugin, and that plugin can (should?) ensure its data is updated to the context as part of init, but only if the stored data is newer than any previously set value.
After init, but before start, the context can be updated to all plugins (via `onContextUpdate`) _if_ it was updated during init.

Note, verification shouldn't be updating the TSS data in most cases; certainly not during init (we intend to remove any storage or configuration for TSS data in the Verification plugin, the ApplicationStateFacility and TSS Bootstrap will handle that instead). The only time Verification might update TSS data is based on Block `0` (and even that could migrate to another plugin if appropriate).

Author: ApplicationStateFacility and all the plugins can now init without getting any context updates. Any TssData updates they submit will be added to a ConcurrentLinkedQueue. After plugin initialization, the ApplicationStateFacility takes the TssData with the greatest validFromBlock and calls the onContextUpdate method for each plugin.

   - Plugins will be updated using the `BlockNodePlugin.onContextUpdate()`.
   - Initial `TssData` will be provided in the BlockNodeContext.
2. Will the `TssBootstrapPlugin` persist the .bin file into the Verification PVC?
   - Plugins should not share configuration information.
   - The verification plugin should manage its own configuration
   - The `ApplicationStateFacility` will be responsible for persisting the `TssData`
   - When the verification plugins detects new `TssData`, ie Block 0, it will notify the `ApplicationStateFacility`.

>    - The `ApplicationStateFacility` will update all plugins of BlockNodeContext changes via `BlockNodePlugin.onContextUpdate()`

Reviewer: This is good, but perhaps add one more line at the top.
   - No plugin will write to storage managed by another plugin.

Author: Added

3. Is this plugin intended to run only once at BN first run.
   - The plugin will periodically query its peer BN servers to see if they have any `TssData` updates.

> 4. Where and how it will get the BN source(s) for trusted peers to get the TSS data from? We already have the concept of backfill peers.

Reviewer: Minor grammar fix:

```suggestion
4. Where and how will it get the BN source(s) for trusted peers to get the TSS data from? We already have the concept of backfill peers.
```

# Design Document Template

## Table of Contents

1. [Purpose](#purpose)
2. [Goals](#goals)
3. [Terms](#terms)
4. [Entities](#entities)
5. [Design](#design)
6. [Diagram](#diagram)
7. [Configuration](#configuration)
8. [Metrics](#metrics)
9. [Exceptions](#exceptions)
10. [Acceptance Tests](#acceptance-tests)

## Purpose

We intend to create a new Block Node plugin (TSSBootstrapPlugin) that queries another Block Node's detailed status to pick up TSS configuration details needed to initiate TSS verification.

## Goals

> - Delivers TSS booststrap information to a Block Node that does not have access to the network Block 0 transactions.

Reviewer 2: nit: quite lengthy goals. 
This reads as a mixture of goals and design specs.

For this I might expect goals  as follows
1. BN detects missing TSS state and queries peer BNs for data
2. BN validate TSS Data on load
3. BN propagates TSS data load/changes to other plugins

The flow of logic and how the entities interact would then be in the Design section

Author: Updated

     */
    uint64 valid_from_block = 4;
}

/**
 * TSS Roster information.
 *
 * `TssRoster` contains a list of roster entry
 */
message TssRoster {
    /**
     * The list of `RosterEntry`
     */
    repeated RosterEntry roster_entries = 1;
    /**
     * The starting block number this TssData is valid from
     */
    uint64 valid_from_block = 2;
}

/**
 * A single TSS Roster entry.
 *
 * All fields are REQUIRED.<br/>
 * The `node_id` field SHALL match the same field for an entry in the Node Store
 * in consensus network state.<br/>
 * The `schnorr_public_key` MAY be a placeholder value if the associated node failed
 * to participate in the roster election within the required time limit.
 */
message RosterEntry {
    /**
     * The node id
     */
    uint64 node_id = 1;
    /**
     * The node weight
     */
    uint64 weight = 2;
    /**
     * The schnorr public key
     */
    bytes schnorr_public_key = 3;
}
```

## Design

- Completed design decisions
  - Should the TssData or TssRoster messages have the starting block number that the TssData is valid from.
    - valid_from_block has been added to both TssData and TssRoster.

## Diagram

> TBD

Reviewer 2: One of these diagrams to show case the flow of the TSS data between entities would be a good illustration

## Diagram

TBD

Consider using mermaid to generate one or more of the following:
- [User Journey](https://mermaid.js.org/syntax/userJourney.html)
- [Requirement Diagram](https://mermaid.js.org/syntax/requirementDiagram.html)
- [Sequence Diagram](https://mermaid.js.org/syntax/sequenceDiagram.html)
- [Class Diagram](https://mermaid.js.org/syntax/classDiagram.html)
- [Block Diagram](https://mermaid.js.org/syntax/block.html)
- [Architecture Diagram](https://mermaid.js.org/syntax/architecture.html)
- [Flowchart](https://mermaid.js.org/syntax/flowchart.html)
- [State Diagram](https://mermaid.js.org/syntax/stateDiagram.html)
- [Entity Relationship Diagram](https://mermaid.js.org/syntax/entityRelationshipDiagram.html)**

## Configuration

### ApplicationStateFacility (NodeConfig)

- `tssDataFilePath`. The file path to use to persist and load `TssData`
- `tssUpdatePollIntervalMillis`. The poll interval in milliseconds that the ApplicationStateFacility uses to check for
  `TssData` updates

### TssBootstrapPlugin (TssBootstrapConfig)

- `ledgerId`. Base64 encoded `ledgerId`
- `wrapsVerificationKey`. Base64 encoded `wrapsVerificationKey`

> - `nodeId`. The node id

Reviewer 2: I saw this in the implementation. 
Why do we only have one entry in the roster covered here?
We can update in follow up
